docs/NNNN-hashed-n-tuple-storage-layout.md

```diff
@@ -0,0 +1,242 @@
+* **Name**: `tupleSize`
+  * **Description:** Indicates the size of the segments (in characters) that the digest is split into
+  * **Type:** integer
+  * **Range:** 0,32
+  * **Default:** 3
+* **Name:** `numberOfTuples`
+  * **Description:** Indicates how many segments are used for path generation
+  * **Type:** integer
+  * **Range:** 0,32
+  * **Default:** 3
+* **Name:** `shortObjectRoot`
+  * **Description:** When true, indicates that the OCFL object root directory name should contain the remainder of the digest not used in the n-tuples segments
+  * **Type:** boolean
+  * **Default:** false
+
+### Details
+
+#### digestAlgorithm
+
+`digestAlgorithm` is defaulted to `sha256`, and it MUST either contain a digest algorithm that's [officially supported by the OCFL spec](https://ocfl.io/draft/spec/#digest-algorithms) or defined in a community extension. The specified algorithm is applied to OCFL object identifiers to produce hex encoded digest values that are then mapped to OCFL object root paths.
+
+#### tupleSize
+
+`tupleSize` determines the number of digest characters to include in each tuple. The tuples are used as directory names. The default value is `3`, which means that each directory in the OCFL storage hierarchy could contain up to 4096 sub-directories. Increasing this value increases the maximum number of sub-directories per directory.
+
+If `tupleSize` is set to `0`, then no tuples are created and `numberOfTuples` MUST also equal `0`.
+
+The product of `tupleSize` and `numberOfTuples` MUST be less than or equal to the number of characters in the hex encoded digest.
+
+#### numberOfTuples
+
+`numberOfTuples` determines how many tuples to create from the digest. The tuples are used as directory names, and each successive directory is nested within the previous. The default value is `3`, which means that every OCFL object root will be 4 directories removed from the OCFL storage root, 3 tuple directories plus 1 encapsulation directory. Increasing this value increases the depth of the OCFL storage hierarchy.
+
+If `numberOfTuples` is set to `0`, then no tuples are created and `tupleSize` MUST also equal `0`.
+
+The product of `numberOfTuples` and `tupleSize` MUST be less than or equal to the number of characters in the hex encoded digest.
+
+#### shortObjectRoot
+
+The directory that immediately encapsulates an OCFL object MUST either be named using the entire digest or the remainder of the digest that was not used in a tuple. When `shortObjectRoot` is set to `false`, the default, the entire digest is used, and, when it's `true` only the previously unused remainder is used.
+
+If the product of `tupleSize` and `numberOfTuples` is equal to the number of characters in the hex encoded digest, then `shortObjectRoot` MUST be `false`.
+
+## Procedure
+
+The following is an outline of the steps to map an OCFL object identifier to an OCFL object root path:
+
+1. The OCFL object identifier is hashed using the specified `digestAlgorithm`.
```

Comment:
should you specify the encoding of the object identifier for hashing? Would it be encoded as UTF8?

Reply:
@bcail I'm happy to add it, but I'm not sure if it needs to be specified. It's interesting to note that the OCFL spec never mentions anything about character encoding. However, my assumption is that the OCFL object ID (and all logical paths) MUST implicitly be utf-8 encoded because it's stored in JSON. The [JSON spec](https://tools.ietf.org/html/rfc8259#section-8) states, "JSON text exchanged between systems that are not part of a closed ecosystem MUST be encoded using UTF-8." I guess there's some ambiguity here as to whether or not an OCFL repository can/should be considered a "closed ecosystem". In which case, it would be nice if the OCFL spec itself clarified.

Reply:
OK, yeah if there's a global assumption that all encodings will be UTF-8, then it doesn't need to be specified here, but it would be nice if it were explicitly said somewhere. Seems like it should be clear what encoding was used to generate the path, without trying various encodings...

Reply:
@bcail Dug this up this morning: https://github.com/OCFL/spec/issues/284

Reply:
It seems like both the encoding of inventory.json, and what a filesystem does with unicode characters in paths, are actually outside of this issue. This is an issue that is completely at the level of the ocfl client, where it's determining what the paths should be.

Is there anything telling me I shouldn't hash the UTF16 bytes of an object ID, to determine what the path to the object will be? Am I allowed to do that? If I can do that, then how does the client know encoding was used, when it's trying to find an object written by a different client? It seems like me like an explicit requirement for UTF8 would remove any questions.

Reply:
I'll change it based on the outcome of [this issue](https://github.com/OCFL/spec/issues/514). If the intention is to support more than just UTF-8, then I don't see a reason to restrict the representation here. In which case, it would likely make sense to have the encoding be a parameter.

Reply:
@bcail Thinking about it some more, you're right that the encoding of the inventory is more or less irrelevant here. I had been thinking that if OCFL supported more than UTF-8 that it might be nice to get the encodings consistent, but I don't think there's necessarily a reason to. I will update here to require UTF-8. Thanks for the feedback!